doc/edgehub-api.md

```diff
@@ -0,0 +1,434 @@
+
+Because of the request/response pattern, IoT Hub/Edge Hub publishes the twin on a topic and the device or module that requested the twin needs to be subscribed to this topic to receive the result. Specificly, the device/module needs to subscribe to the following topic:
+```
+$iothub/twin/res/#
+```
+When a device/module wants to get the latest twin, it needs to generate a request id, which is an arbitrary string, like a guid or an incremented counter. Then it needs to publish an empty message on the following topic:
+```
+$iothub/twin/GET/?$rid={request id}
+```
+When Iot Hub/Edge Hub responds to the twin request, it sends back 3 things:
+- a status code for the operation, which is a similar number to HTTP status codes
+- the request id the client used to request the twin
+- the twin itself
+
+The status code is usually 200, which means successful operation. The code 429 is returned in case of too many twin request, when IoT Hub throttled the request. Status codes starting with 5** are other errors.
+
+An example response topic for successful operation and request id = 1234:
+```
+$iothub/twin/res/200/?$rid=1234
+```
+The mqtt message payload is an utf8 encoded json document showed earlier describing twin messages,
+
+When using Edge Hub, the twin resquest will be handled the following way:
+- When a device/module connects to Edge Hub, it pulls the latest twin from IoT Hub, so it has a copy
+- When a device/module sends a twin request, Edge Hub forwards the request to Iot Hub and waits for the answer
+- When a response arrives from Iot Hub, Edge Hub relays the answer to the device/module.
+- If there is an error requesting the twin from IoT Hub, Edge Hub sends its latest copy to the device/module instead.
+### Sending Reported Property Updates
+Reported Property Updates are request/response pairs similarly to twin pulls. The request/response is necessary because the update can fail for different reason and the sender needs to know to repeat it.
+
+Just like for twin pulls, the sender needs to generate a request id, which will be sent back by IoT Hub/Edge Hub. Then the update message must be published on the following topic:
+```
+ $iothub/twin/PATCH/properties/reported/?$rid={request id}
+```
+The payload of the update message describes all the properties with the new values:
+```
+{
+    "UpdatedProperty1" : "new value",
+    "UpdatedProperty2" : 123,
+    "UpdatedProperty3" : null
+}
+```
+In the example three properties are getting updated, all the rest (if any) of the reported properies will keep their old values. The third "UpdatedProperty3" makes the propery deleted, as it is described in the [documentation of device twins](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins).
+
+The response is going to be published on a topic like the following:
+```
+$iothub/twin/res/204/?$rid=1234
```

Comment:
Since the topic is the same for Property Update and Getting Twin, How does a device know how to parse the response? Is it based on what request id is returned?

Reply:
Yes, the way it works is that the logic that sent a property update, it generated a unique request id (rid=xyz), so it waits for that response which contains rid-xyz and continues from there. As you are saying, because the topic is the same for two operations, it does not work that there is a generic receiver which then dispatches the incoming packet for "twin result handling" or "update result handling", because it cannot know which it is. Always the sender of the original message needs to monitor if it got back a response.